Asynch UART

robamu opened the pull request.

robamu committed:
Added asynchronous UART TX driver

jonathanpallant commented on example-code/qemu-aarch32v8r/src/bin/global_uart.rs:
```diff
@@ -8,7 +8,7 @@
 
 use core::fmt::Write;
-use qemu_aarch32v8r::uart::{CmsdkUart, MutexUart, UART0_ADDR};
+use qemu_aarch32v8r::uart::{Uart, MutexUart, UART0_ADDR};
```
What was the reason for renaming `CmsdkUart` to just `Uart`? We have other examples in this folder which use a PL011 UART, which requires a different driver, so I don't want to lose the context here about which UART driver is which.

robamu replied:
I think the module was named `cmsdk_uart`, so I considered the CMSDK naming in the UART structure redundant. But I can revert the naming, so this is explicit in contexts where CMSDK is not in the module name.

robamu replied:
I reverted the naming.

jonathanpallant commented on example-code/qemu-common/src/cmsdk_uart/asynch.rs:
```diff
@@ -0,0 +1,178 @@
+#[derive(Debug)]
+pub struct InvalidWakerIndex(pub usize);
+
+pub fn on_interrupt_tx(tx: &mut Tx, uart_index: usize) {
```
What is the purpose of `uart_index`? If it's just to handle multiple async drivers (each handling a different UART), then maybe a static AtomicUsize that initialises to zero and we increment every time `TxAsynch::new()` is called, might be better?

robamu replied:
Implemented with `TxHandlerId` now.

jonathanpallant commented on example-code/qemu-common/src/cmsdk_uart/asynch.rs:
```diff
@@ -0,0 +1,178 @@
+
+impl embedded_io_async::Write for TxAsynch {
+    /// It is very important that the data which is sent outlives the asynchronous futures
+    /// created with it.
```
Can we enforce this with lifetime annotations?

robamu replied:
I am not even sure if this is necessary and maybe the comment can be removed. If I create a future for the asynchronous write, instead of awaiting it, the data is borrowed and can not be modified.

jonathanpallant commented on example-code/qemu-thumbv7em/src/bin/uart_asynch.rs:
```diff
@@ -0,0 +1,67 @@
+
+        let mut index = 0;
+        loop {
+            defmt::info!("Writing string {} asycnhronously", index);
```
```suggestion
            defmt::info!("Writing string {} asynchronously", index);
```

robamu commented on example-code/qemu-common/src/cmsdk_uart/asynch.rs:
```diff
@@ -0,0 +1,178 @@
+use core::{cell::RefCell, convert::Infallible, sync::atomic::AtomicBool};
```
maybe we should add `#[deny(missing_docs)]` to the lib

jonathanpallant committed:
Refactored the async UART example.

robamu committed:
small renaming

jonathanpallant committed:
Simplify the UART APIs.

The joint UART type does everything through its Rx and Tx components. Except for enabling interrupts, where it's more efficient to do both together.

Also ensures a critical section is used when a shared register is modified (because Tx and Rx could modify it concurrently).

Overflow interrupts are disabled because we weren't usinfg them.

robamu commented on example-code/qemu-common/src/cmsdk_uart/basic.rs:
```diff
@@ -76,72 +81,74 @@ impl CmsdkUart {
-            c.set_rxie(true);
-            c
-        });
+    pub fn regs(&mut self) -> &mut MmioRegisters<'static> {
```
merged it

jonathanpallant commented:
I appreciate you sticking with this. It's a great addition to the material ... let's goooooooooo!
